src/main/java/org/dcsa/ebl/service/impl/TransportDocumentTOServiceImpl.java

```diff
@@ -50,70 +52,123 @@
+                                        .flatMap(shippingInstructionTO -> {
+                                            transportDocumentTO.setShippingInstruction(shippingInstructionTO);
+                                            String carrierBookingReference = shippingInstructionTOService.getCarrierBookingReference(shippingInstructionTO);
+                                            if (carrierBookingReference == null) {
+                                                return Mono.error(new IllegalStateException("No CarrierBookingReference specified on ShippingInstruction:" + shippingInstructionTO.getId() + " - internal error!"));
+                                            } else {
+                                                return Flux.concat(
+                                                        updateTransportDocumentWithBookingInfo(carrierBookingReference, transportDocumentTO),
+                                                        createCharges(transportDocumentTO, shippingInstructionTO.getIsChargesDisplayed()),
+                                                        updateTransportDocumentWithTransportPlan(carrierBookingReference, transportDocumentTO)
+                                                ).then();
+                                            }
+                                        })
                                         .thenReturn(transportDocumentTO),
-                                createCharges(transportDocumentTO),
-                                createClauses(transportDocumentTO),
-                                createTransportPlan(transportDocumentTO)
+                                // Create charges if any
+                                // Create/Update clauses if any
+                                createClauses(transportDocumentTO)
                         )
                         .then(Mono.just(transportDocumentTO));
                     });
         }
     }
 
-    private Flux<Charge> createCharges(TransportDocumentTO transportDocumentTO) {
+    /**
+     * Checks if TransportDocument differs from Booking with regards to ServiceType, ShipmentTerms and ContractService.
+     * Sets the above mentioned values from Booking on TransportDocument
+     * @param carrierBookingReference the reference to the Booking
+     * @param transportDocumentTO the TransportDocument to update
+     * @return Error if a discrepancy is found otherwise returns the Booking
+     */
+    private Mono<Booking> updateTransportDocumentWithBookingInfo(String carrierBookingReference, TransportDocumentTO transportDocumentTO) {
+        // Check if TransportDocument differs from values in Booking
+        return getBooking(carrierBookingReference, transportDocumentTO.getShippingInstructionID())
+                .flatMap(booking -> {
+                    if (transportDocumentTO.getServiceTypeAtOrigin() != null && !transportDocumentTO.getServiceTypeAtOrigin().equals(booking.getServiceTypeAtOrigin())) {
```

Comment:
Can this be shorter if we use `Objects.equals`

(applies to the following if-statements)